src/arkode/arkode_lsrkstep.c

```diff
@@ -0,0 +1,2220 @@
+
+  The output variable dsmPtr should contain estimate of the
+  weighted local error if an embedding is present; otherwise it
+  should be 0.
+
+  The input/output variable nflagPtr is generally used in ARKODE 
+  to gauge the convergence of any algebraic solvers. However, since 
+  the STS step routines do not involve an algebraic solve, this variable 
+  instead serves to identify possible ARK_RETRY_STEP returns within this 
+  routine.
+
+  The return value from this routine is:
+            0 => step completed successfully
+           >0 => step encountered recoverable failure;
+                 reduce step and retry (if possible)
+                 ARK_RETRY_STEP indicates that the required stage
+                 number has reached the stage_max_limit with the
+                 current value of h. The step is then returned to
+                 adjust the step size.
+           <0 => step encountered unrecoverable failure
+  ---------------------------------------------------------------*/
+int lsrkStep_TakeStepRKC(ARKodeMem ark_mem, sunrealtype* dsmPtr, int* nflagPtr)
+{
+  int retval;
+  sunrealtype hmax, w0, w1, temp1, temp2, arg, bjm1, bjm2, mus, thjm1, thjm2,
+    zjm1, zjm2, dzjm1, dzjm2, d2zjm1, d2zjm2, zj, dzj, d2zj, bj, ajm1, mu, nu,
+    thj;
+  const sunrealtype onep54 = SUN_RCONST(1.54), c13 = SUN_RCONST(13.0),
+                    p8 = SUN_RCONST(0.8), p4 = SUN_RCONST(0.4);
+  ARKodeLSRKStepMem step_mem;
+
+  /* initialize algebraic solver convergence flag to success,
+     temporal error estimate to zero */
+  *nflagPtr = ARK_SUCCESS;
+  *dsmPtr   = ZERO;
+
+  /* access ARKodeLSRKStepMem structure */
+  retval = lsrkStep_AccessStepMem(ark_mem, __func__, &step_mem);
+  if (retval != ARK_SUCCESS) { return retval; }
+
+  sunrealtype* cvals = step_mem->cvals;
+  N_Vector* Xvecs    = step_mem->Xvecs;
+
+  /* Compute dominant eigenvalue and update stats */
+  if (step_mem->dom_eig_update)
+  {
+    retval = lsrkStep_ComputeNewDomEig(ark_mem, step_mem);
+    if (retval != ARK_SUCCESS) { return retval; }
+  }
+
+  sunrealtype ss =
+    SUNRceil(SUNRsqrt(onep54 * SUNRabs(ark_mem->h) * step_mem->spectral_radius));
+  ss = SUNMAX(ss, SUN_RCONST(2.0));
+
+  if (ss >= step_mem->stage_max_limit)
+  {
+    if (!ark_mem->fixedstep)
+    {
+      hmax = ark_mem->hadapt_mem->safety * SUNSQR(step_mem->stage_max_limit) /
+             (onep54 * step_mem->spectral_radius);
+      ark_mem->eta = hmax / ark_mem->h;
+      *nflagPtr    = ARK_RETRY_STEP;
+      ark_mem->hadapt_mem->nst_exp++;
+      return ARK_RETRY_STEP;
+    }
+    else
+    {
+      arkProcessError(ark_mem, ARK_MAX_STAGE_LIMIT_FAIL, __LINE__, __func__,
+                      __FILE__,
+                      "Unable to achieve stable results: Either reduce the "
+                      "step size or increase the stage_max_limit");
+      return ARK_MAX_STAGE_LIMIT_FAIL;
+    }
+  }
+
+  step_mem->req_stages = (int)SUNRround(ss);
```

Comment:
Bringing this up again since I'm not sure we settled on a solution, but the potential undefined behavior from the cast is still present. I advocate for
```suggestion
  step_mem->req_stages = SUNMIN(ss, INT_MAX);
```
which handles edges cases like overflows and NANs correctly. See https://godbolt.org/z/q5qx7f3s7

Either way, the `SUNRround` is unnecessary since `ss` cannot have a fractional part (it's the result of `SUNRceil`).

Reply:
This might fail the CI because of an overflow warning, see https://godbolt.org/z/b8v5xqGz6

Reply:
This is a lot of discussion for an issue that cannot occur -- just a few lines up @maggul explicitly handled the case where `ss` exceeded `stage_max_limit` (which is an `int`).  Thus, if that block of code did not occur (which would `return` out of the function), then it is impossible for `ss` to exceed the representable limit in this line.  If we want to unnecessarily calculate `SUNMIN` then by all means we can keep this suggestion in the code, but I'd prefer that it go back to the more efficient (and perfectly correct) version that @maggul already had.

Reply:
Ah ok, I missed the `stage_max_limit` check. So the `SUNMIN` shouldn't be needed (as well as the `SUNRround`)

Reply:
I mentioned this check on the other branch, but I should have mentioned it here as well to ensure you've seen it. Nonetheless, I believe the last commit should resolve the issue. Please let me know if you think there's still room for improvement. I'd appreciate it if you could review the other areas sooner rather than later.